File: cc/025.cc
#include <iostream>
#include <vector>
using namespace std;

class TreeNode {
public:
    int val;
    TreeNode * left;
    TreeNode * right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
    vector<vector<int> > paths;
    vector<int> path;
public:
    vector<vector<int> > FindPath(TreeNode* root, int expectNumber) 
    {
        if (root)
        { 
            path.push_back(root->val);
            expectNumber -= root->val;
            if (!expectNumber && !root->left && !root->right)
            {
                vector<int> copy(this->path);
                this->paths.push_back(copy);
            }
            this->FindPath(root->left, expectNumber);
            this->FindPath(root->right, expectNumber);
            this->path.pop_back();
            expectNumber += root->val;
        }
        return this->paths;
    }
};

int main(int argc, char const *argv[])
{
    TreeNode * pRoot = new TreeNode(0);
    pRoot->left = new TreeNode(3);
    pRoot->left->left = new TreeNode(2);
    pRoot->left->right = new TreeNode(6);
    pRoot->right = new TreeNode(4);
    pRoot->right->left = new TreeNode(5);
    pRoot->right->right = new TreeNode(1);
     /*
        0
       / \
      3   4
     / \ / \
     2 6 5 1
     */
    (new Solution())->FindPath(pRoot, 9);
    return 0;
}

File: python/025.py
# -*- coding: utf-8 -*-
# /usr/bin/python3
# Creator: Hemajun
# Description: 输入一颗二叉树和一个整数，打印出二叉树中结点值的和为输入整数的所有路径。路径定义为从树的根结点开始往下一直到叶结点所经过的结点形成一条路径。


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def __init__(self):
        self.__paths = []
        self.__path = []

    # 返回二维列表，内部每个列表表示找到的路径
    def FindPath(self, root, expectNumber):
        if root:
            self.__path.append(root.val)
            expectNumber -= root.val
            if expectNumber == 0 and root.left == None and root.right == None:
                self.__paths.append(self.__path[:])
            self.FindPath(root.left, expectNumber)
            self.FindPath(root.right, expectNumber)
            self.__path.pop()
            expectNumber += root.val
        return self.__paths
